packages/elasticsearch-client/src/search.ts

```diff
@@ -0,0 +1,145 @@
+/**
+ * Copyright (C) 2021-2023 Technology Matters
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as published
+ * by the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see https://www.gnu.org/licenses/.
+ */
+
+import {
+  SearchParameters,
+  SearchQuery,
+  SearchQueryFilters,
+  SearchResults,
+} from '@tech-matters/types';
+
+import { getClient } from './client';
+import getConfig from './get-config';
+
+import getAccountSid from './get-account-sid';
+
+interface SearchTotalHits {
+  value: number;
+  relation: 'eq' | 'gte';
+}
+
+/**
+ * f track_total_hits is false, Elasticsearch returns an approximate count of the total
+ * hits as a number in the total field. If track_total_hits is true, Elasticsearch returns
+ * an object of type SearchTotalHits that provides more accurate information about the total
+ * hits.
+ */
+const getTotalValue = (total: number | SearchTotalHits | undefined): number => {
+  if (typeof total === 'object') return total.value;
+
+  return total || 0;
+};
+
+// TODO: this doesn't support range filters yet
+export const generateFilters = (filters: SearchParameters['filters']): SearchQueryFilters => {
+  const returnFilters: SearchQueryFilters = [];
+
+  if (!filters?.length) return returnFilters;
+
+  Object.entries(filters).forEach(([key, value]) => {
+    if (Array.isArray(value)) {
+      returnFilters.push({
+        terms: {
+          [key]: value,
+        },
+      });
+    } else {
+      returnFilters.push({
+        term: {
+          [key as string]: value,
+        },
+      });
+    }
+  });
+
+  return returnFilters;
+};
+
+export const generateElasticsearchQuery = (
+  accountSid: string,
+  searchParameters: SearchParameters,
+  fields: string[],
+) => {
+  const { q, filters, pagination } = searchParameters;
+  const { limit, start } = pagination;
+
+  const query: SearchQuery = {
+    index: `${accountSid.toLowerCase()}-resources`,
+    body: {
+      query: {
+        bool: {
+          must: [
+            {
+              query_string: {
+                query: q,
+                fields,
+              },
+            },
+          ],
+        },
+      },
+      from: start,
+      size: limit,
+    },
+  };
+
+  if (filters) {
+    query.body.query.bool.filter = generateFilters(filters);
```

Comment:
I'm not sure if this method already expects the parameters to have already been sanitised at this point, just want to point out that a 'filter' from the API isn't always expected to be transformed to a 'filter' for the search index (provided the concept of 'filter' in ElasticSearch and CloudSearch are analogous), some of them might be interpreted as search terms.

For example, 'province' is a UI filter, but we should probably use it as a search term because we want to look for it across several fields, not just a specific field. We could have the UI decide which filters are really filters and which ones are search terms, but I thought it would be better if how each part of the search was implemented was a black box to the client

Reply:
I'm assuming that once we add in some real filters we'll move the `generateFilters()` function and possibly even the entire generateElasticsearchQuery() down into the per index search config and customize it quite a bit more. This was meant as more of an example of how we would set keyword based filters eventually than a fully fleshed out filter system.